src/compiler/parser.ts

```diff
@@ -3480,7 +3518,7 @@ module ts {
                 case SyntaxKind.VoidKeyword:
                     return parseVoidExpression();
                 case SyntaxKind.LessThanToken:
-                    return parseTypeAssertion();
+                    return (sourceFile.jsxFactory && tryParse(() => parseJSXElement(true))) || parseTypeAssertion();

```

Comment:
CAn you give examples where you get unfriendly errors?

Even if we're only parsing JSX elmeents in .tsx files (or in the presence of this directive) it may still be very impactful on perf to do this.  We would still want people to be able to edit large .tsx files in the IDE without parsing being super expensive.

I'm also worried about the parsing cost because this lookahead is exponentially bad. i.e. you can be speculatively parsing one element, then you have to do that again inside the element.  Then if that outer element fails to parse, then we do that work again when we hit the inner element.